### PR TITLE
Feature dynamic sidecar

### PR DESCRIPTION
**Pull request by erdrix**
Add the following features into the CassandraCluster definition: 

- Dynamic PVC definition
- Dynamic Containers (= Sidecars) definition

### CODE REVIEW

File: pkg/apis/db/v1alpha1/register.go [hunk 1]
```diff
@@ -12,8 +12,11 @@ import (
 
 var (
 	// SchemeGroupVersion is group version used to register these objects
-	SchemeGroupVersion = schema.GroupVersion{Group: "db.orange.com", Version: "v1alpha1"}
+	GroupVersion = schema.GroupVersion{Group: "db.orange.com", Version: "v1alpha1"}
```
Comment (cscetbon) [inline pkg/apis/db/v1alpha1/register.go]: Shouldn't it be alpha2 here ?

File: pkg/controller/cassandracluster/generator.go [hunk 2]
```diff
@@ -182,6 +182,36 @@ func generateCassandraVolumeMount(cc *api.CassandraCluster) []v1.VolumeMount {
...
+	}
+
+	return vms
+}
+
+
+func generateStorageConfigVolumeClaimTemplates(cc *api.CassandraCluster, labels map[string]string) []v1.PersistentVolumeClaim {
+
+	var pvcs []v1.PersistentVolumeClaim
+
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/generator.go]: let's remove one line here

File: pkg/controller/cassandracluster/generator.go [hunk 3]
```diff
@@ -182,6 +182,36 @@ func generateCassandraVolumeMount(cc *api.CassandraCluster) []v1.VolumeMount {
...
+
+	for _, storage := range cc.Spec.StorageConfigs {
+		var pvc v1.PersistentVolumeClaim
+		pvc = v1.PersistentVolumeClaim{
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/generator.go]: can just do `pvc := v1.PersistentVolumeClaim..` here and remove `var pvc ...`

File: deploy/role.yaml [hunk 4]
```diff
@@ -36,25 +38,25 @@ rules:
...
+  - '*'
 - apiGroups:
   - monitoring.coreos.com
   resources:
   - servicemonitors
   verbs:
-  - "get"
-  - "create"
+  - get
+  - create
```
Comment (cscetbon) [inline deploy/role.yaml]: 👍

File: pkg/apis/db/v1alpha1/register.go (hunk 1 above)
Comment (cscetbon) [reply pkg/apis/db/v1alpha1/register.go]: nvm

File: pkg/apis/db/v1alpha1/zz_generated.deepcopy.go [hunk 5]
```diff
@@ -110,6 +110,20 @@ func (in *CassandraClusterSpec) DeepCopyInto(out *CassandraClusterSpec) {
...
+		in, out := &in.StorageConfigs, &out.StorageConfigs
+		*out = make([]StorageConfig, len(*in))
+		for i := range *in {
+			(*in)[i].DeepCopyInto(&(*out)[i])
+		}
+	}
+	if in.SidecarConfigs != nil {
+		in, out := &in.SidecarConfigs, &out.SidecarConfigs
+		*out = make([]v1.Container, len(*in))
+		for i := range *in {
```
Comment (cscetbon) [inline pkg/apis/db/v1alpha1/zz_generated.deepcopy.go]: I'd use something else than `i` like `index` or `idx`

Comment (cscetbon) [reply pkg/apis/db/v1alpha1/zz_generated.deepcopy.go]: wait is it generated code ? if yes, nvm

Comment (allamand) [reply pkg/apis/db/v1alpha1/zz_generated.deepcopy.go]: Yes this is generated ^^

Comment (allamand) [reply pkg/apis/db/v1alpha1/zz_generated.deepcopy.go]: Yes this is generated ^^

File: deploy/role.yaml [hunk 6]
```diff
@@ -1,14 +1,16 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: Role
 metadata:
+  creationTimestamp: null
```
Comment (cscetbon) [inline deploy/role.yaml]: not needed

File: pkg/controller/cassandracluster/testdata/cassandracluster-2DC.yaml [hunk 7]
```diff
@@ -25,6 +25,54 @@ spec:
...
+          name: cassandra-logs
+    - args: ["tail", "-F", "/var/log/cassandra/gc.log.0.current"]
+      image: ez123/alpine-tini
+      imagePullPolicy: Always
+      name: gc-logs
+      resources:
+        limits:
+          cpu: 50m
+          memory: 50Mi
+        requests:
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/testdata/cassandracluster-2DC.yaml]: let's use anchors like in other files as much as we can

File: pkg/controller/cassandracluster/testdata/cassandracluster-2DC.yaml [hunk 8]
```diff
@@ -25,6 +25,54 @@ spec:
...
+          cpu: 10m
+          memory: 10Mi
+      volumeMounts:
+        - mountPath: /var/log/cassandra
+          name: cassandra-logs
+    - args: ["tail", "-F", "/var/log/cassandra/gc.log.0.current"]
+      image: ez123/alpine-tini
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/testdata/cassandracluster-2DC.yaml]: why not just alpine ? the size is the same

File: pkg/controller/cassandracluster/generator_test.go [hunk 9]
```diff
@@ -130,22 +132,178 @@ func TestGenerateCassandraService(t *testing.T) {
...
+	return false
+}
+
+/**
+ Generate volume list
+ */
+func generateCassandraVolumeMounts() []v1.VolumeMount {
+	var vms []v1.VolumeMount
+	vms = append(vms, v1.VolumeMount{Name: "data", MountPath: "/var/lib/cassandra"})
+	//vms = append(vms, v1.VolumeMount{Name: "cassandra-config", MountPath: "/configmap"})
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/generator_test.go]: comments to remove

File: pkg/controller/cassandracluster/generator_test.go [hunk 10]
```diff
@@ -130,22 +132,178 @@ func TestGenerateCassandraService(t *testing.T) {
 func TestGenerateCassandraStatefulSet(t *testing.T) {
 	assert := assert.New(t)
 
-	_, cc := helperInitCluster(t, "cassandracluster-2DC.yaml")
-	labels, nodeSelector := k8s.GetDCRackLabelsAndNodeSelectorForStatefulSet(cc, 0, 0)
-	sts := generateCassandraStatefulSet(cc, &cc.Status, "dc1", "dc1-rack1", labels, nodeSelector, nil)
-
-	assert.Equal(map[string]string{
+	expectedLabels := map[string]string{
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/generator_test.go]: It's a matter of taste I guess. I usually do not use variables when it's used once

File: deploy/role.yaml (hunk 6 above)
Comment (erdrix) [reply deploy/role.yaml]: It's generated by the operator-sdk !

**Pull request merged**